### ISSUE TITLE
incorrect initial loss for softmax loss/classifier

### ISSUE THREAD

**Issue by jkarimi91** [fchollet/keras#6259]
In a properly configured network, the expected initial loss should be ~log(num_classes) = ~2.3, however, instead we have a log of ~14; what is causing this discrepancy?

```python
from cs231n.data_utils import get_CIFAR10_data
from keras.models import Sequential
from keras.layers import Conv2D, MaxPool2D, Dense, Flatten
from keras.optimizers import Adam
from keras.utils import to_categorical


data = get_CIFAR10_data()
for k, v in data.iteritems():
  print '%s: ' % k, v.shape

#X_val:  (1000, 3, 32, 32)
#X_train:  (49000, 3, 32, 32)
#X_test:  (1000, 3, 32, 32)
#y_val:  (1000,)
#y_train:  (49000,)
#y_test:  (1000,)

X_val, y_val = data['X_val'], data['y_val']
X_train, y_train = data['X_train'], data['y_train']
X_test, y_test = data['X_test'], data['y_test']

y_train = to_categorical(y_train, num_classes=10)
y_val = to_categorical(y_val, num_classes=10)

model = Sequential()
model.add(Conv2D(32, 3, padding='same', activation='relu', input_shape=(3, 32, 32),
                 data_format='channels_first'))
model.add(Conv2D(32, 3, padding='same', activation='relu', data_format='channels_first'))
model.add(MaxPool2D(data_format='channels_first'))

model.add(Conv2D(32, 3, padding='same', activation='relu', data_format='channels_first'))
model.add(Conv2D(32, 3, padding='same', activation='relu', data_format='channels_first'))
model.add(MaxPool2D(data_format='channels_first'))

model.add(Flatten())
model.add(Dense(500, activation='relu'))
model.add(Dense(10, activation='softmax'))

optim = Adam(lr=1e-3)
model.compile(loss='categorical_crossentropy',
              optimizer=optim,
              metrics=['accuracy'])

model.fit(X_train, y_train, batch_size=100, epochs=1, validation_data=(X_val, y_val))
```

**Comment by jkarimi91** [comment]
[stackoverflow post](http://stackoverflow.com/questions/43422164/keras-incorrect-initial-loss-for-softmax-loss-classifier)

**Comment by joelthchao** [comment]
Try to do normalization?
```
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
```
BTW, this is not an issue suitable to post here.

**Comment by jkarimi91** [comment]
@joelthchao good to know, I will make sure to not post issues like this here again. Also, the data is zero-centered

**Comment by joelthchao** [comment]
If you do the normalization I mention in previous reply, initial loss will be around 2.3. With only zero centered, loss starts from ~1x.

**Comment by jkarimi91** [comment]
normalizing by means of scaling worked; I am just trying to figure out why scaling wasn't necessary in my non-keras implementation before I post an answer on SO. Thank you @joelthchao for the help

**Comment by joelthchao** [comment]
Maybe non-keras implementation use other weights initializer (keras uses `glorot_uniform` for kernel).

**Comment by jkarimi91** [comment]
why would weight scales effect the unregularized loss when the softmax function converts the raw scores into normalized probabilities
